### PR TITLE
`VideoInput` operator for the CPU

### PR DESCRIPTION
**Pull request by szalpal**
Signed-off-by: szalpal <mszolucha@nvidia.com>

## Category:

**New feature** (*non-breaking change which adds functionality*)

## Description:
This PR adds a new operator: `VideoInput` operator.

This effort has been preceded with a couple of previous PRs:
#4501 
#4505 
#4508 
#4513 
Some of the changes here are a result of changes introduced in previous PRs. It's really hard to introduce such a big change in one go.

The description of `VideoInput` operator is included in the Operator docstring.

Notable things that happened:
* Implementing `VideoInput` for CPU.
* Adding `inputs` tests to the `python-readers-decoders` suite.
* Relaxing `feed_input` check, to check the presence of the callback only in the `external_source`, not all InputOperators.
* Polishing `feed_input` documentation.
* Relaxing the Uniform Input/Output BatchSize check, not to cover the `VideoInput` operator.

The `VideoInput` is an operator, that might return various batch sizes throughout operation. While `max_batch_size` is constant throughout the pipeline lifetime, the `VideoInput` might return the partial or incomplete batch at the end of the video file. This happens, when the number of frames in the video is insufficient to fill the batch completely. For this reason, the check for a uniform batch size had to be relaxed.

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [x] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

File: dali/operators/input/video_input.cc [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    Invalidate();
+  }
+}
+
+
+DALI_SCHEMA(experimental__inputs__Video)
+                .DocStr(
+                        R"code(
+Decodes a video from a memory buffer. Returns a batch of sequences of frames, where
+each frame has ``(F, H, W, C)`` layout (where ``F`` is the number of frames in a sequence).
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:81]: shouldn't it be
```each frame has (H, W, C) layout```
?
I'd rephrase to:
```
Returns a batch of sequences of frames, with layout ``(F, H, W, C)``, where ``F`` is the number of frames in the sequence. 
```

File: dali/operators/input/video_input.cc [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+When using ``fn.inputs.video`` operator inside the DALI Pipeline, user needs to provide the data
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:83]:
```suggestion
When using ``fn.inputs.video`` operator inside the DALI Pipeline, the user needs to provide the data
```

File: dali/operators/input/video_input.cc [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+using :meth:`Pipeline.feed_input`. When the Operator is fed with data, the Pipeline can be ran
+multiple times and the ``fn.inputs.video`` operator will return consecutive sequences, as long as
+there is enough data remaining. When the source of the frames (the video file) depletes, user needs
+to call another ``feed_input`` to provide the next video file to the operator. This Operator has an
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:87]:
```suggestion
to call ``feed_input`` again to provide the next video file to the operator. This Operator has an
```

File: dali/operators/input/video_input.cc [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+inner-queue for the data, so the ``feed_input`` may be called multiple times and when given video
+file ends, the Operator will fetch the next one automatically from the top of the queue.
+Running the pipeline while there is no data for the ``fn.inputs.video`` to run results in an error.
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:90]: Is there a mechanism (exception based or otherwise) to tell the user we ran out of data without erroring?

**Comment by szalpal** [comment]
!build

File: dali/operators/input/video_input.cc [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+This operator takes only one video as and input (i.e. ``input_batch_size=1``) and will return
+batches of sequences, where every output batch will have the size, which has been set during
+the Pipeline object instantiation (``max_batch_size``). When the number of frames in the video
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:94]:
```suggestion
batches of sequences, where every output batch containing ``max_batch_size``, set during Pipeline creation. When the number of frames in the video
```

File: dali/operators/input/video_input.cc [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+
+The difference between ``fn.inputs.video`` and ``fn.readers.video`` is that the former
+reads an encoded video from memory and the latter reads the encoded video from disk.
+
+The difference between ``fn.inputs.video`` and ``fn.decoders.video`` is that the former
+does not decode the whole video file in one go. This behaviour is needed for longer videos. E.g.
+5-min, 4k, 30fps decoded video takes about 1.7 TB of memory.
+
+This operator accepts most of the video containers and file formats. FFmpeg is used to parse
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:123]: This statement is too vague. Can we list the supported container and formats instead?

File: dali/operators/input/video_input.cc [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+the video container. In the situations, that the container does not contain required metadata
+(e.g. frames sizes, number of frames, etc...), the operator needs to find it out itself,
+which may result in a slowdown.
+)code")
+                .NumInput(0)
+                .NumOutput(1)
+                .AddArg("frames_per_sequence", R"code(
+How many frames shall make up a single sequence returned by the operator.
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:131]: Nitpick: Slightly convoluted language. I'd prefer "Number of frames in each sequence"

File: dali/operators/input/video_input.cc [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+)code", DALI_INT32)
+                .AddOptionalArg("last_sequence_policy", R"code(
+Specifies, how to handle the last sequence in the video file.
+
+For a given number of frames in the video file and ``frames_per_sequence`` parameter,
+it might happen that the video can't be split uniformly across sequences. If the
+``last_sequence_policy='partial'``, the last sequence will have less frames than
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:138]:
```suggestion
``last_sequence_policy='partial'``, the last sequence might have fewer frames than
```

File: dali/operators/input/video_input.cc [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+``frames_per_sequence`` value specified. If the ``last_sequence_policy='partial'``,
+the last sequence will have ``frames_per_sequence`` frames and will be padded with empty frames.
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:140]:
```suggestion
the last sequence will always have ``frames_per_sequence`` frames and will be padded with empty frames.
```

File: dali/operators/decoder/video/video_decoder_base.h [hunk 10]
```diff
@@ -123,6 +123,13 @@ class DLL_PUBLIC VideoDecoderBase {
 
 
   std::vector<std::unique_ptr<FramesDecoder>> frames_decoders_;
+
+ private:
+  using storage_backend_for_copy_kernel = std::conditional_t<
+          std::is_same_v<OutBackend, CPUBackend>,
+          StorageCPU,
+          StorageGPU
+  >;
```
Comment (jantonguirao) [inline dali/operators/decoder/video/video_decoder_base.h:132]: There's already `backend_to_storage_device`. Would be better to use existing tools

Comment (jantonguirao) [reply dali/operators/decoder/video/video_decoder_base.h:132]:
```
using OutStorage = backend_to_storage_device<OutBackend>;
```
^^ I think `OutStorage` is simpler.

File: dali/operators/input/video_input.h [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+ * @param num_frames How many frames the video has.
+ * @param frames_per_sequence How many frames per sequence user requested.
+ * @param batch_size How many sequences make up a single batch.
+ * @return
+ */
+auto DetermineBatchOutline(int num_frames, int frames_per_sequence, int batch_size) {
+  assert(frames_per_sequence > 0);
+  assert(batch_size > 0);
+  // Initially, the Operator will return full batches of full sequences.
+  const int num_full_batches = num_frames / frames_per_sequence / batch_size;
```
Comment (jantonguirao) [inline dali/operators/input/video_input.h:52]:
```suggestion
  const int num_full_batches = num_frames / (frames_per_sequence * batch_size);
```

File: dali/operators/input/video_input.h [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+  explicit VideoInput(const OpSpec &spec) :
+          InputOperator<Backend>(spec),
+          frames_per_sequence_(spec.GetArgument<int>("frames_per_sequence")),
+          device_id_(spec.GetArgument<int>("device_id")),
+          batch_size_(spec.GetArgument<int>("max_batch_size")),
+          last_sequence_policy_(spec.GetArgument<std::string>("last_sequence_policy")) {
+    DALI_ENFORCE(last_sequence_policy_ == "partial" || last_sequence_policy_ == "pad",
+                 make_string("Provided `last_sequence_policy` is not supported: ",
+                             last_sequence_policy_));
+    Invalidate();
```
Comment (jantonguirao) [inline dali/operators/input/video_input.h:90]: This is not necessary here

File: dali/operators/input/video_input.cc (hunk 4 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:90]: There's not. I don't know if we need one. I might be wrong, but I believe that when you're using ExternalSource with `feed_input`, so do not have such mechanism either, is that correct? If so and if we need it, we might design something generic.

For sure, adding such a thing would be rather a bigger effort. ExternalSource has the `cycle` parameter, but the whole operator is partially implemented in Python, therefore it's easier to throw a `StopIteration` there. With `VideoInput` I'm afraid it won't be that simple.

File: dali/operators/decoder/video/video_decoder_base.h (hunk 10 above)
Comment (szalpal) [reply dali/operators/decoder/video/video_decoder_base.h:132]: The problem is, that the `kernels::copy` implementation assumes using `StorageCPU`, `StorageGPU`, etc. Therefore there are the following problems:
1. Plain `OutStorage` cannot be used, because it's `...Backend` not `Storage...`.
1. `backend_to_storage_device` does not provide any information about `Storage...`. It returns `StorageDevice` enum, which is something different and not compatible with `kernels::copy` implementation.

I didn't implement any generic `StorageDevice -> Storage...` mapping, since I didn't want to introduce another feature in this PR, its big enough already.

File: dali/python/nvidia/dali/pipeline.py [hunk 13]
```diff
@@ -878,42 +878,43 @@ def _feed_input(self, name, data, layout=None, cuda_stream=None, use_copy_kernel
         else:
             self._pipe.SetExternalTLInput(name, data, ctypes.c_void_p(cuda_stream), use_copy_kernel)
 
+
     def feed_input(self, data_node, data, layout=None, cuda_stream=None, use_copy_kernel=False):
         """Pass a multidimensional array or DLPack (or a list thereof) to an output of
-           ExternalSource.
+           InputOperator.
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:883]: Done

File: qa/TL0_python-self-test-readers-decoders-inputs/test_body.sh [hunk 14]
```diff
@@ -39,6 +39,9 @@ test_py_with_framework() {
 
     
     ${python_new_invoke_test} -s decoder
+
+
+    ${python_new_invoke_test} -s input
```
[the remark this answers is not in the document]
Comment (szalpal) [reply qa/TL0_python-self-test-readers-decoders-inputs/test_body.sh:44]: Good point. Reverted the directory rename. Hopefully that'll be ok, if `test-readers-decoders` will test also `inputs`?

File: dali/operators/input/video_input.cc (hunk 1 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:81]: How about:
```
Returns a batch of sequences of frames, 
with the layout: ``(N, F, H, W, C)``, where:

* ``N`` - number of sequences in a batch,
* ``F`` - number of frames in a sequence,
* ``H`` - height of the frame,
* ``W`` - width of the frame,
* ``C`` - number of channels in the frame.```

File: dali/operators/input/video_input.cc (hunk 2 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:83]: Done

File: dali/operators/input/video_input.cc (hunk 3 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:87]: Done

File: dali/operators/input/video_input.cc (hunk 7 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:131]: Done

File: dali/operators/input/video_input.cc (hunk 8 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:138]: Done

File: dali/operators/input/video_input.cc (hunk 9 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:140]: Done

File: dali/operators/input/video_input.h (hunk 12 above)
Comment (szalpal) [reply dali/operators/input/video_input.h:90]: Done

File: dali/test/python/input/test_video.py [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    decoder_pipe.feed_input(input_name, [test_file])
+    decoder_out = decoder_pipe.run()
+
+    input_pipe.build()
+    input_pipe.feed_input(input_name, np.array([[test_file]]))
+
+    for ref_seq in portion_out_reference_sequence(decoder_out, frames_per_sequence, batch_size):
+        input_out = input_pipe.run()
+        test_seq = input_out[0].as_array()
+        assert np.all(ref_seq == test_seq)
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/test/python/input/test_video.py:134]: Done

File: dali/operators/input/video_input.cc (hunk 6 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:123]: I believe it might be problematic, because there are dozens of those.

FTR, I copy-pasted this sentence from `readers.video` documetation: https://docs.nvidia.com/deeplearning/dali/main-user-guide/docs/operations/nvidia.dali.fn.readers.video.html

File: dali/operators/input/video_input.cc (hunk 5 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:94]: Is this OK:
```
This operator takes only one video as and input (i.e. ``input_batch_size=1``) and will return
batches of sequences. Every output batch will have the ``max_batch_size``, set during the Pipeline
creation.```

File: dali/operators/input/CMakeLists.txt [hunk 16]
```diff
@@ -14,5 +14,13 @@
...
+
+if (BUILD_FFMPEG)
+    list(APPEND DALI_INST_HDRS "${CMAKE_CURRENT_SOURCE_DIR}/video_input.h")
+    list(APPEND DALI_OPERATOR_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/video_input.cc")
+endif (BUILD_FFMPEG)
+
+set(DALI_OPERATOR_SRCS ${DALI_OPERATOR_SRCS} PARENT_SCOPE)
+set(DALI_INST_HDRS ${DALI_INST_HDRS} PARENT_SCOPE)
+
+collect_test_sources(DALI_OPERATOR_TEST_SRCS PARENT_SCOPE)
```
Comment (stiepan) [inline dali/operators/input/CMakeLists.txt:26]: Will that work if `BUILD_FFMPEG` is OFF?

Comment (stiepan) [reply dali/operators/input/CMakeLists.txt:26]: Wouldn't `input/video` and conditional adding of a whole video dir be cleaner?

File: dali/operators/input/video_input.cc [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+  }
+}
+
+
+DALI_SCHEMA(experimental__inputs__Video)
+                .DocStr(
+                        R"code(
+Streams and decodes a video from a memory buffer. To be used with long and high resolution videos.
+
+Returns a batch of sequences of frames, with the layout: ``(N, F, H, W, C)``, where:
```
Comment (stiepan) [inline dali/operators/input/video_input.cc:83]: Do you actually set the layout to the output tensor?

File: dali/operators/input/video_input.h [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+ * Determines, how is the video file split into batches.
+ *
+ * For a given video file and frames_per_sequence and batch_size parameters, we can determine
+ * beforehand, how do the return sequences look like in terms of their length:
+ *
+ * For instance, for num_frames=67, frames_per_sequence=5 and batch_size=3:
+ * --------------- --------------- --------------- --------------- -------
+ * [   ][   ][   ] [   ][   ][   ] [   ][   ][   ] [   ][   ][   ] [   ][]
+ * --------------- --------------- --------------- --------------- -------
+ * 4 full batches of full sequences, and 1 (last) partial batch with 2 full sequences
```
Comment (stiepan) [inline dali/operators/input/video_input.h:40]:
```suggestion
 * 4 full batches of full sequences, and 1 (last) partial batch with 1 full sequence
```
?

File: dali/operators/input/video_input.h [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+    od.shape = uniform_list_shape(
+            num_full_sequences + (frames_in_last_sequence == 0 ? 0 : 1),
+            GetSequenceShape(frames_per_sequence_, 0));
+
+    if (frames_in_last_sequence != 0) {
+      // And at the end of the last batch, the last sequence will be either partial or padded,
+      // depending on user setting.
+      od.shape.set_tensor_shape(od.shape.num_samples() - 1, GetSequenceShape(
+              last_sequence_policy_ == "pad" ? frames_per_sequence_ : frames_in_last_sequence, 0));
+    }
```
Comment (stiepan) [inline dali/operators/input/video_input.h:149]:
```suggestion
    // And at the end of the last batch, the last sequence will be either partial or padded,
    // depending on user setting.
    if (frames_in_last_sequence != 0 && last_sequence_policy_ != "pad") {
      od.shape.set_tensor_shape(od.shape.num_samples() - 1, GetSequenceShape(frames_in_last_sequence, 0));
    }
```
?

File: dali/operators/input/video_input.cc [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+  for (int64_t s = 0; s < output.num_samples(); s++) {
+    auto pad_value =
+            last_sequence_policy_ == "pad" ? std::optional<SampleView<CPUBackend>>(GetPadFrame())
+                                           : std::nullopt;
+    full_sequence = DecodeFrames(output[s], 0, frames_per_sequence_, pad_value);
+    if (!full_sequence) {
+      break;
+    }
+  }
+  if (!full_sequence || frames_decoders_[0]->NextFrameIdx() == -1) {
```
Comment (stiepan) [inline dali/operators/input/video_input.cc:72]: Is this possible that we have `!full_sequence` but `frames_decoders_[0]->NextFrameIdx() > 0`? What happens then?

File: dali/operators/input/video_input.cc [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+Returns a batch of sequences of frames, with the layout: ``(N, F, H, W, C)``, where:
+
+* ``N`` - number of sequences in a batch. Currently ``N=max_batch_size``,
+* ``F`` - number of frames in a sequence,
+* ``H`` - height of the frame,
+* ``W`` - width of the frame,
+* ``C`` - number of channels in the frame.
+
+When using ``fn.inputs.video`` operator inside the DALI Pipeline, the user needs to provide the data
+using :meth:`Pipeline.feed_input`. When the Operator is fed with data, the Pipeline can be ran
```
Comment (stiepan) [inline dali/operators/input/video_input.cc:92]:
```suggestion
using :meth:`Pipeline.feed_input`. When the Operator is fed with data, the Pipeline can be run
```

File: dali/operators/input/video_input.cc [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+multiple times and the ``fn.inputs.video`` operator will return consecutive sequences, as long as
+there is enough data remaining. When the source of the frames (the video file) depletes, user needs
```
Comment (stiepan) [inline dali/operators/input/video_input.cc:94]: how about

```suggestion
there is enough data to decode. When the source of the frames (the video file) depletes, user needs
```

File: dali/operators/input/CMakeLists.txt (hunk 16 above)
Comment (szalpal) [reply dali/operators/input/CMakeLists.txt:26]: This directory will also have a GPU variant with `BUILD_NVDEC` option, so I believe it's better with particular `if`.

Fixed the `collect_test_sources` part.

File: dali/operators/input/video_input.h (hunk 18 above)
Comment (szalpal) [reply dali/operators/input/video_input.h:40]: That's correct.

File: dali/operators/input/video_input.cc (hunk 17 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:83]: I'd prefer to leave the description of the batch in the layout too, unless @jantonguirao thinks it shouldn't be here?

Added the test for the layout in `test_video.py`.

File: dali/python/nvidia/dali/pipeline.py [hunk 23]
```diff
@@ -879,41 +879,43 @@ def _feed_input(self, name, data, layout=None, cuda_stream=None, use_copy_kernel
             self._pipe.SetExternalTLInput(name, data, ctypes.c_void_p(cuda_stream), use_copy_kernel)
 
     def feed_input(self, data_node, data, layout=None, cuda_stream=None, use_copy_kernel=False):
-        """Pass a multidimensional array or DLPack (or a list thereof) to an output of
-           ExternalSource.
+        """Pass a multidimensional array or DLPack (or a list thereof) to the eligible Operator.
```
Comment (stiepan) [inline dali/python/nvidia/dali/pipeline.py:882]:
```suggestion
        """Pass a multidimensional array or DLPack (or a list thereof) to the eligible operator.
```

It is just a nitpick, but for me the capitialized operator reads weird here (like a term from the first section of super formal agreement where the definiton of parties is introduced).

I think input operator instead of InputOperators would be easier to read. For example you write `The operators that`, but in the same sentence `are the InputOperators`. Wouldn't simply operators, operator and input operators be enough? When referring to actual code you use `fn.inputs` and `fn.external_source`, which should be enough.

File: dali/operators/input/video_input.cc [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+5-min, 4k, 30fps decoded video takes about 1.7 TB of memory.
+
+This operator accepts most of the video containers and file formats. FFmpeg is used to parse
+the video container. In the situations, that the container does not contain required metadata
+(e.g. frames sizes, number of frames, etc...), the operator needs to find it out itself,
+which may result in a slowdown.
+)code")
+                .NumInput(0)
+                .NumOutput(1)
+                .AddArg("frames_per_sequence", R"code(
```
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/operators/input/video_input.cc:145]: True. On the other hand, I feel that `frames_per_sequence` is better, since `sequence_length` may be in seconds and we have a request for a variable frame rate videos to alternatively specify duration. I'll ask other guys in the team.

File: dali/operators/input/video_input.cc (hunk 20 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:72]: My intention was that if by any chance `ReadNextFrame` fails (making the sequence partial), but `frames_decoders_[0]->NextFrameIdx()` still return `>0`, that's when we want to invalidate the stream anyhow. Unless you feel it's excessive check?

File: dali/operators/input/video_input.cc (hunk 21 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:92]: Shouldn't it be a past form here? :)

Like

> "When the Operator is fed with data and the Pipeline can be..."

> "...assigned as peculiar"
> "...multiplied by 42"
> "...covered with snow"

File: dali/test/python/input/test_video.py [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+    pad_pipe = video_input_pipeline(input_name=input_name, batch_size=batch_size,
+                                    frames_per_sequence=frames_per_sequence, device=device,
+                                    last_sequence_policy='pad', **common_pipeline_params)
+
+    num_frames = get_num_frames(test_video)
+
+    partial_pipe.build()
+    partial_pipe.feed_input(input_name, np.array([[test_video]]))
+    pad_pipe.build()
+    pad_pipe.feed_input(input_name, np.array([[test_video]]))
```
Comment (stiepan) [inline dali/test/python/input/test_video.py:152]: Do you ever test scheduling multiple files in the queue through number of feed_input calls?

File: dali/operators/input/video_input.cc (hunk 21 above)
Comment (stiepan) [reply dali/operators/input/video_input.cc:92]: Yep, but it `can be done` not `can be did`. :D

File: dali/operators/input/video_input.cc (hunk 17 above)
Comment (stiepan) [reply dali/operators/input/video_input.cc:83]: What worries me is that the TL layout (that is printed in python with `__repr__` will say FHWC), which may be confusing. If you were to pass the output to another dali python you would specify the layout as "FHWC" too, right? Instead of N I would just mention the number of samples/sequences in the batch. Having said that, I am not going to insist, just leaving my concerncs here.

File: dali/test/python/input/test_video.py (hunk 25 above)
Comment (szalpal) [reply dali/test/python/input/test_video.py:152]: Good point. Added the test.

File: dali/operators/decoder/video/video_decoder_base.h (hunk 10 above)
Comment (jantonguirao) [reply dali/operators/decoder/video/video_decoder_base.h:132]: My bad, the one I meant is `storage_tag_map`, from views.h

File: dali/operators/input/video_input.cc (hunk 17 above)
Comment (jantonguirao) [reply dali/operators/input/video_input.cc:83]: Actually, leaving the batch information in the layout is a bit misleading, since we omit it purposedly in the rest of DALI documentation. We've seen situations where the user tried to use the batch information in the layout and had to explain that in DALI the batch is implicit and we do not handle batch information explicitly in the pipeline.. So yes, I agree with Kamil that the batch information should be omitted here.

File: dali/operators/input/video_input.cc (hunk 24 above)
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/operators/input/video_input.cc:145]: My vote would go to sticking to `sequence_length` as well.

File: dali/operators/input/video_input.cc (hunk 21 above)
Comment (jantonguirao) [reply dali/operators/input/video_input.cc:92]: Yeah, it's run :)

File: dali/operators/input/video_input.cc [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+    encoded_videos.set_pinned(device_id_ != CPU_ONLY_DEVICE_ID);
+    frames_decoders_.resize(input_batch_size);
+    auto &thread_pool = ws.GetThreadPool();
+    this->ForwardCurrentData(encoded_videos, thread_pool);
+
+    // Creating FramesDecoders
+    auto sample = encoded_videos[0];
+    auto data = reinterpret_cast<const char *>(sample.data<uint8_t>());
+    size_t size = sample.shape().num_elements();
+    frames_decoders_[0] = std::make_unique<FramesDecoder>(data, size, false);
```
Comment (jantonguirao) [inline dali/operators/input/video_input.cc:40]: FramesDecoderGpu has a different signature than FramesDecoder:
```
  FramesDecoderGpu(
    const char *memory_file,
    int memory_file_size,
    cudaStream_t stream = 0,
    bool build_index = true,
    int num_frames = -1);
```
vs.
```
  FramesDecoder(const char *memory_file, int memory_file_size, bool build_index = true,
                bool init_codecs = true, int num_frames = -1);
```.

Wouldn't `false` be interpreted as stream in the GPU case?

File: dali/operators/input/video_input.cc (hunk 4 above)
Comment (jantonguirao) [reply dali/operators/input/video_input.cc:98]: So, does the user need to know how many times we can call run until the next feed_input needs to be called? I was just wondering how the usage will look in practice

File: dali/operators/input/video_input.cc (hunk 17 above)
Comment (jantonguirao) [reply dali/operators/input/video_input.cc:83]: BTW, I can see that readers.video has the batch dimension in the description. I prefer if we stick to omitting the layout, otherwise the users will start fetching/setting/etc the channel dimension and will be confused. However, you may choose to leave it as is to be consistent with video reader and I won't be insisting more.

File: dali/operators/input/video_input.cc (hunk 26 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:40]: That's correct, but this is a `VideoInput<CPUBackend>::SetupImpl`. I'll be writing `VideoInput<GPUBackend>::SetupImpl` for GPU operator.

File: dali/operators/decoder/video/video_decoder_base.h (hunk 10 above)
Comment (szalpal) [reply dali/operators/decoder/video/video_decoder_base.h:132]: Unfortunately, `storage_tag_map` cannot be used. For the GPU operator there will not be `VideoInput<GPUBackend>`, but `VideoInput<MixedBackend>`. And the problem is, that a mapping `MixedBackend -> StorageGPU` shouldn't be introduced as a general one.

File: dali/operators/input/video_input.cc (hunk 21 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:92]: Oh, that's right ;)

File: dali/operators/input/video_input.cc (hunk 20 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:72]: Oh, actually you had it correct in the first comment. `NextFrameIdx` used in this context tells, if there are any more frames within the `FramesDecoder`. I'll add a comment, since it's not clear.

The information, that there are no more frames in the `FramesDecoder` is vital for the mechanism, that will be introduced in the next PR. I.e. even when we decoded a full sequence, but there are no more frames in the decoder, we have to tell the user to call the `feed_input` again.

File: dali/operators/input/video_input.cc (hunk 4 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:98]: In the C API we will have the `EventId` mechanism (name TBD). This is the way user will know when to call `feed_input`. In the Python API, it is not decided yet. I mean, it may and probably will be the `EventId` mechanism also, but this is a different and big effort to actually add it to the Python API. For now we're assuming that user knows how many frames the video has and therefore knows when to call `feed_input`. That's, BTW, how I wrote the tests.

File: dali/operators/input/video_input.h (hunk 19 above)
Comment (szalpal) [reply dali/operators/input/video_input.h:149]: Done

File: dali/python/nvidia/dali/pipeline.py (hunk 23 above)
Comment (szalpal) [reply dali/python/nvidia/dali/pipeline.py:882]: Fixed

File: dali/operators/input/video_input.cc (hunk 17 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:83]: I went over everything again, thought it through again and @stiepan 's point makes it. When specifying the layout in another operator, user would put `FHWC`, not `NFHWC`. I'll omit the `N` in the documentation.

File: dali/test/python/input/test_video.py (hunk 25 above)
Comment (szalpal) [reply dali/test/python/input/test_video.py:152]: Good point. I'll add a test

File: dali/operators/input/video_input.cc [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+multiple times and the ``fn.inputs.video`` operator will return consecutive sequences, as long as
+there is enough data to decode. When the source of the frames (the video file) depletes, user needs
+to call another ``feed_input`` again to provide the next video file to the operator. This Operator
+has an inner-queue for the data, so the ``feed_input`` may be called multiple times and when given
+video file ends, the Operator will fetch the next one automatically from the top of the queue.
+Running the pipeline while there is no data for the ``fn.inputs.video`` to run results in an error.
+
+This operator takes only one video as and input (i.e. ``input_batch_size=1``) and will return
+batches of sequences. Every output batch will have the ``max_batch_size``, set during the Pipeline
+creation. When the number of frames in the video file does not allow to split the frames uniformly
```
Comment (stiepan) [inline dali/operators/input/video_input.cc:102]:
```suggestion
batches of sequences. Every output batch will consist of ``max_batch_size`` samples (the ``max_batch_size` is set during the Pipeline creation). When the number of frames in the video file does not allow to split the frames uniformly
```

The exact wording in not that important, but if I understand what you mean here, the word `samples` is missing, right?

File: dali/operators/input/video_input.cc (hunk 20 above)
Comment (stiepan) [reply dali/operators/input/video_input.cc:73]: Okay! A comment for a future readers would defienitely be helpful.

File: dali/operators/input/video_input.cc (hunk 27 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:102]: I intended `max_batch_size size`, but samples works also good 👍🏻

File: dali/operators/input/video_input.cc (hunk 24 above)
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/operators/input/video_input.cc]: Changed name to `sequence_length`

File: dali/operators/input/video_input.cc (hunk 1 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc]: I removed the batch dimension from the doc.

**Comment by szalpal** [comment]
!build

**Comment by szalpal** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

File: dali/operators/input/video_input.cc (hunk 20 above)
Comment (szalpal) [reply dali/operators/input/video_input.cc:73]: If you don't mind, I'll add it in a followup PR. Wouldn't like to rerun CI for just a comment ;)

File: dali/operators/input/video_input.cc [hunk 28]
```diff
@@ -0,0 +1,166 @@
...
+)code", DALI_INT32)
+                .AddOptionalArg("last_sequence_policy", R"code(
+Specifies, how to handle the last sequence in the video file.
+
+For a given number of frames in the video file and ``frames_per_sequence`` parameter,
+it might happen that the video can't be split uniformly across sequences. If the
+``last_sequence_policy='partial'``, the last sequence might have fewer frames than
+``frames_per_sequence`` value specified. If the ``last_sequence_policy='partial'``,
+the last sequence will always have ``frames_per_sequence`` frames and will
+be padded with empty frames.
```
Comment (stiepan) [inline dali/operators/input/video_input.cc:157]:
```suggestion
For a given number of frames in the video file and ``sequence_length`` parameter,
it might happen that the video can't be split uniformly across sequences. If the
``last_sequence_policy='partial'``, the last sequence might have fewer frames than
``sequence_length`` value specified. If the ``last_sequence_policy='partial'``,
the last sequence will always have ``sequence_length`` frames and will
be padded with empty frames.
```

Leftover.

**Review by stiepan — APPROVED** [review]
